### PR TITLE
Fix --steps-scaler so short runs keep the full training curriculum

### PR DESCRIPTION
**Pull request by Neilus03**
## Summary

Fix `--steps-scaler` so it scales every step-dependent training setting and resolves automatic defaults from the final scaled step count.

## Why I looked at this

While testing shorter versions of the Lego training run on an Apple Silicon Mac, I compared:

```bash
caffeinate -i .venv/bin/metal-gauss-train \
  --blender data/nerf_synthetic/lego \
  --steps 30000 \
  --steps-scaler 0.1 \
  --max-resolution 800 \
  --num-downscales 2 \
  --eval-every 3000 \
  --seed 0
```

On a clean checkout of upstream `main`, this command did run for 3,000 effective steps, but the training report showed that several automatic defaults had been selected from the original 30,000-step value:

- `budget`: 500,000
- `resolution_schedule`: 10,000
- `start_active`: 150,000

As a result, the 3,000-step run never reached the scheduled resolution changes and stayed at the coarsest 200×200 resolution for its entire training run. It also allocated a 500,000-Gaussian capacity intended for a much longer run.

There was no exception. The command appeared to complete successfully, so this was easy to miss when looking only at the training logs.

I reproduced this from clean upstream source with no PR-specific changes. The issue is in `metal_gauss/train.py`: it derived the default resolution schedule and automatic Gaussian budget before applying `--steps-scaler`.

This is therefore a training-configuration and reproducibility issue in Metal-Gauss itself, rather than a Metal kernel, MPS, dataset, or machine-specific failure.

## What changed

- Resolve `--steps-scaler` in the dependency-free schedule module before deriving automatic defaults.
- Scale the effective step count, relocation interval, evaluation interval, SH warmup, resolution interval, 3D-filter refresh interval, and checkpoint-export interval together.
- Keep zero sentinels disabled for SH warmup, 3D-filter refresh, and checkpoint export.
- Keep an explicitly supplied `--budget` unchanged, since it is a Gaussian-capacity override rather than a time-dependent schedule.
- Preserve the existing `start_active` safety clamp after the final Gaussian budget is known.
- Reject non-positive or non-finite scaler values with a clear command-line error.
- Add CPU-only regression tests for:
  - default resolution derivation;
  - automatic budget derivation;
  - explicit step-dependent intervals;
  - disabled zero sentinels;
  - invalid scaler values;
  - benchmark-report validation.

## Results

The table below compares the same user-facing command before and after the fix. Both runs use 3,000 effective training steps.

The clean upstream implementation derives its defaults from the unscaled 30,000-step value, while the patched implementation derives them from the final 3,000-step value.

| Run | Resolved training setup | Held-out PSNR | Training wall time |
| --- | --- | ---: | ---: |
| Clean upstream `main` at `25bf135` | 3,000 steps; 500,000 Gaussians; 200×200 throughout | 20.91 dB | 158.2 s |
| Patched commit `7302a38` | 3,000 steps; 100,000 Gaussians; 200×200 → 400×400 → 800×800 | 28.48 dB | 73.3 s |

The patched run reached 7.57 dB higher held-out PSNR and completed in approximately half the training time in this comparison.

These numbers should not be interpreted as a new Metal-kernel optimization. The improvement comes from correcting the training protocol so that a scaled run actually uses a schedule appropriate for its shorter effective duration.

## Qualitative comparison

The following figure compares the same three official Lego test cameras rendered at 512×512 with a white background.

The two reconstructed columns use the same cameras, background, and Metal rendering path. The only difference between them is whether the clean upstream implementation or this patch was used.

Qualitative comparison of upstream and patched runs

_Figure 1. From left to right, each row shows the reference image, the clean upstream run at commit `25bf135`, and the patched run at commit `7302a38`. The rows use official test views 0, 50, and 100. The per-view PSNR values are shown below the rendered outputs._

The patched render is visibly sharper and contains less blur and ghosting around the Lego structure, roof, wheels, and base. This provides a qualitative view of the same improvement reported by the held-out evaluation numbers above.

[Open the full-resolution comparison image](https://github.com/Neilus03/metal-gauss/blob/336871e814a99cf1db0869dd0fdc8c746b1a67b6/outputs/steps-scaler-qualitative-comparison.png)

## Resolved configuration after the fix

For the command above, the patched implementation resolves the scaled training schedule as follows:

| Setting | Resolved value |
| --- | ---: |
| Effective steps | 3,000 |
| Automatically selected Gaussian budget | 100,000 |
| Initial active Gaussians | 50,000 |
| Relocation interval | 10 steps |
| Evaluation interval | 300 steps |
| SH warmup | 100 steps |
| Resolution transition interval | 1,000 steps |
| 3D-filter refresh interval | 0, disabled |
| Checkpoint-export interval | 0, disabled |

The important difference is that the automatic budget and resolution schedule are now derived after the effective step count has been scaled.

## Validation

The following checks passed:

- `.venv/bin/python -m pytest -q` — 178 passed
- `.venv/bin/python bench/readme_tables.py --check` — passed
- `.venv/bin/python -m py_compile metal_gauss/schedule.py metal_gauss/train.py bench/runner.py tests/test_schedule.py tests/test_runner.py` — passed
- `git diff --check` — passed
- Clean Apple Silicon/MPS run at commit `7302a38` — completed successfully with `dirty=false` in the trainer report

## Apple Silicon environment

The Metal training and rendering runs were performed on:

- Apple Silicon Mac
- macOS 26.6.2
- Python 3.12.14
- PyTorch 2.13.0 with MPS

## Scope

This is a focused correctness and reproducibility fix for the existing `--steps-scaler` option.

It does not change the renderer, introduce a new rasterization algorithm, or make a new Metal-kernel performance claim. It makes short scaled training runs follow the schedule that their effective step count implies.

### CODE REVIEW

**Comment by Neilus03** [comment]
Thanks for catching both of these.

The first issue was valid: the benchmark harness was allowing any mismatch for step-dependent fields whenever `steps_scaler` was non-unit. I updated the transform check so it now calculates the exact expected scaled value and only accepts the divergence when the trainer reports that value. The zero-sentinel behavior for disabled intervals is also checked explicitly.

I added regression coverage for incorrect values across the step count, relocation interval, evaluation interval, SH warmup, resolution schedule, 3D-filter refresh interval, and checkpoint-export interval.

The second issue was also valid: `tests/test_schedule.py` was not included in the CPU-only GitHub Actions job. I added it to the existing dependency-free test invocation. The schedule tests run without torch, a GPU, or the Metal backend.

The updated checks pass locally with 188 tests, and the GitHub Actions run for commit `5541c12` is green.

**Comment by nandometzger** [comment]
I've pushed one commit to your branch removing outputs/steps-scaler-qualitative-comparison.png — 888 KB that git would carry forever, and the image is already in this thread. Merging/Rebasing now :)

**Comment by Neilus03** [comment]
> I've pushed one commit to your branch removing outputs/steps-scaler-qualitative-comparison.png — 888 KB that git would carry forever, and the image is already in this thread. Merging/Rebasing now :)
> 

Yep, probably shouldnt have pushed the png, srry abt that. I wanted to show the difference qualitatively.
